src/autoscaler/api/policyvalidator/policy_validator.go

```diff
@@ -0,0 +1,369 @@
+}
+
+type PolicyValidationError struct {
+	gojsonschema.ResultErrorFields
+}
+
+type DateTimeRange struct {
+	startDateTime time.Time
+	endDateTime   time.Time
+}
+
+func newDateTimeRange(startDateTime string, endDateTime string, timezone string) *DateTimeRange {
+	location, _ := time.LoadLocation(timezone)
+	dateTimeRange := DateTimeRange{}
+	dateTimeRange.startDateTime, _ = time.ParseInLocation(DateTimeLayout, startDateTime, location)
+	dateTimeRange.endDateTime, _ = time.ParseInLocation(DateTimeLayout, endDateTime, location)
+	return &dateTimeRange
+}
+
+func (dtr *DateTimeRange) overlaps(otherDtr *DateTimeRange) bool {
+	if otherDtr.endDateTime.Sub(dtr.startDateTime) > 0 && dtr.endDateTime.Sub(otherDtr.startDateTime) > 0 {
+		return true
+	}
+	return false
+}
+
+func newPolicyValidationError(context *gojsonschema.JsonContext, formatString string, errDetails gojsonschema.ErrorDetails) *PolicyValidationError {
+	err := PolicyValidationError{}
+	err.SetType("custom_invalid_policy_error")
+	err.SetContext(context)
+	err.SetDescriptionFormat(formatString)
+	err.SetDetails(errDetails)
+	return &err
+}
+
+func NewPolicyValidator(policySchemaPath string) *PolicyValidator {
+	policyValidator := &PolicyValidator{
+		policySchemaPath: policySchemaPath,
+	}
+	policyValidator.policySchemaLoader = gojsonschema.NewReferenceLoader("file://" + policyValidator.policySchemaPath)
+	return policyValidator
+}
+
+func (pv *PolicyValidator) ValidatePolicy(policyStr string) error {
+	policyLoader := gojsonschema.NewStringLoader(policyStr)
+
+	result, err := gojsonschema.Validate(pv.policySchemaLoader, policyLoader)
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+
+	if !result.Valid() {
+		return getErrorWithJSONMessage(result.Errors())
+	}
+
+	policy := models.ScalingPolicy{}
+	err = json.Unmarshal([]byte(policyStr), &policy)
+	if err != nil {
+		return err
+	}
+
+	pv.validateAttributes(&policy, result)
+
+	if len(result.Errors()) > 0 {
+		return getErrorWithJSONMessage(result.Errors())
+	}
+	return nil
+}
+
+func (pv *PolicyValidator) validateAttributes(policy *models.ScalingPolicy, result *gojsonschema.Result) {
+
+	rootContext := gojsonschema.NewJsonContext("(root)", nil)
+
+	//check InstanceMinCount and InstanceMaxCount
+	if policy.InstanceMin >= policy.InstanceMax {
+		instanceMinContext := gojsonschema.NewJsonContext("instance_min_count", rootContext)
+		errDetails := gojsonschema.ErrorDetails{
+			"instance_min_count": policy.InstanceMin,
+			"instance_max_count": policy.InstanceMax,
+		}
+		formatString := "instance_min_count {{.instance_min_count}} is higher or equal to instance_max_count {{.instance_max_count}}"
+		err := newPolicyValidationError(instanceMinContext, formatString, errDetails)
+		result.AddError(err, errDetails)
+	}
+
+	// check InstanceMinCount, IntanceMaxCount and InitialInstanceMinCount for schedules
+	if policy.Schedules == nil {
+		return
+	}
+	schedulesContext := gojsonschema.NewJsonContext("schedules", rootContext)
```

Comment:
Missing the implementation for threshold validation as
https://github.com/cloudfoundry-incubator/app-autoscaler/blob/master/api/lib/validation/attributeValidator.js#L238

Reply:
Added